funnel/models/label.py

```diff
@@ -1,67 +1,167 @@
+
+    #: Sequence number for this label, used in UI for ordering
+    seq = db.Column(db.Integer, nullable=False)
+
+    # A single-line description of this label, shown when picking labels (optional)
+    description = db.Column(db.UnicodeText, nullable=False, default=u"")
+
+    #: Icon for displaying in space-constrained UI. Contains one emoji symbol.
+    #: Since emoji can be composed from multiple symbols, there is no length
+    #: limit imposed here
+    icon_emoji = db.Column(db.UnicodeText, nullable=True)
+
+    #: Restricted mode specifies that this label may only be applied by someone with
+    #: an editorial role (TODO: name the role). If this label is a parent, it applies
+    #: to all its children
+    _restricted = db.Column('restricted', db.Boolean, nullable=False, default=False)
+
+    #: Required mode signals to UI that if this label is a parent, one of its
+    #: children must be mandatorily applied to the proposal. The value of this
+    #: field must be ignored if the label is not a parent
+    _required = db.Column('required', db.Boolean, nullable=False, default=False)
+
+    #: Archived mode specifies that the label is no longer available for use
+    #: although all the previous records will stay in database.
+    _archived = db.Column('archived', db.Boolean, nullable=False, default=False)
+
+    #: Proposals that this label is attached to
+    proposals = db.relationship(Proposal, secondary=proposal_label, lazy='dynamic', backref='labels')
+
+    __table_args__ = (db.UniqueConstraint('project_id', 'name'),)
+
+    __roles__ = {
+        'all': {
+            'read': {
+                'name', 'title', 'project_id', 'project', 'seq',
+                'restricted', 'required', 'archived'
+            }
+        }
+    }
+
+    @hybrid_property
+    def restricted(self):
+        return self.parent_label._restricted if self.parent_label else self._restricted
+
+    @restricted.setter
+    def restricted(self, value):
+        if self.parent_label:
+            raise ValueError("Cannot restrict a child label")
+        self._restricted = value
+
+    @restricted.expression
+    def restricted(cls):
+        return case([
+            (cls.parent_label_id != None, db.select([Label._restricted]).where(Label.id == cls.parent_label_id).as_scalar())  # NOQA
+        ], else_=cls._restricted)
+
+    @hybrid_property
+    def archived(self):
+        return self._archived or self.parent_label._archived if self.parent_label else False
+
+    @archived.setter
+    def archived(self, value):
+        if self.parent_label:
+            raise ValueError("Cannot archive a child label")
+        self._archived = value
+
+    @archived.expression
+    def archived(cls):
+        return case([
+            (cls.parent_label_id != None, db.select([Label._archived]).where(Label.id == cls.parent_label_id).as_scalar())  # NOQA
+        ], else_=cls._archived)
```

Comment:
This is incorrect. Child labels can be archived, and that archived flag takes priority.